File: docs/collatz.js
/*
 *  A tree of connected nodes
 *
 *  Each node has an index >= 0
 *  and an odd value of n = 2*index+1
 *  and a step towards its successor.
 *
 *  A succ always exists.
 *  pred and sib are calculated on demand.
 *
 *  relations are:
 *
 *  (2 * succ.index +1) * 2**step == 3 * index + 2;
 *
 *  sib.index = 4 * index + 2;
 *
 */

function Tree() {

    function Node(succ, step) {
        this.step = step;
        this.max = this;

        if(succ) {
            this.succ = succ;
            this.len = succ.len+1;
            this.index = ((2*succ.index+1) * (1<<step) - 2) / 3;
            this.path = succ.path.concat(".").concat(step);
            this.sum = succ.sum + step;
            if(succ.max.index>this.index)
                this.max = succ.max;
            this._fac = undefined;
        } else {
            // root node(0)
            this.succ = this;
            this._pred = this;
            this.len = 0;
            this.index = 0;
            this.path = "1";
            this.sum = 1;
            this._fac = 1;
        }

        this.n = 2*this.index+1;

        this.pred = function(i) {

            const mod = this.index%3;

            if(mod===1)
                return undefined;

            if(i>0)
                return this.pred(i-1).sibl();

            let pred = this._pred;
            if(!pred) {
                const step = mod ? 0 : 1;
                pred = new Node(this, step);
                this._pred = pred;
            }

            return pred;
        }

        this.sibl = function() {
            let sibl = this._sibl;
            if(!sibl) {
                sibl = new Node(this.succ, this.step+2);
                this._sibl = sibl;
            }
            return sibl;
        }

        this.fac = function() {
            let fac = this._fac;
            if(!fac) {
                fac = 1;
                for(let i=1; i<this.index; i++) {
                    if(this.contains(i))
                        ++fac;
                }
                this._fac = fac;
            }
            return fac;
        }

        this.contains = function(i) {
            i = 2*i+1;
            do {

                if(i===this.n)
                    return true;

                i = (3*i+1)/2;
                while((i%2)===0)
                    i = i/2;

            } while(i>1)

            return false;
        }

        this.succeeds = function(node) {
            
            if(node===undefined)
                return false;

            if(node===this)
                return true;

            if(node===this.index)
                return true;

            if(this.succ===this)
                return false;

            return this.succ.succeeds(node);
        }

        this.toString = function() {
            return this.n.toString().concat('(').concat(this.sum).concat(':').concat(this.path).concat(')');
        }
    }

    this.root = new Node(undefined, 1);

    this.node = function(index) {

        if(index===0)
            return this.root;

        let k = 3 * index + 2;
        let step = 0;
        while(k>1 && !(k%2)) {
            ++step;
            k>>=1;
        }

        // to index
        k = (k-1)/2;

        step >>= 1;

        return this.node(k).pred(step);
    }

    this.inode = function(index) {

        if(index === undefined)
            return null;

        if(index%2!==1)  // even
            return null;
            
        return this.node((index-1)/2);
    }
}

File: collatz/src/main/java/de/ditz/collatz/Collatz.java
package de.ditz.collatz;

import java.math.BigInteger;
import java.util.AbstractList;
import java.util.Comparator;
import java.util.List;

/**
 * Created by IntelliJ IDEA.
 * User: stueken
 * Date: 10.04.20
 * Time: 17:43
 */

/**
 * Class Collatz represents an odd entry with n = 2k + 1.
 * Each entry has a known successor towards k=0 (n=1) which is some steps away.
 * Each entry may have a predecessor.
 *
 * Entries with k%3==1 have no predecessor.
 * Entries with k%3==0 have even predecessors (0, 2, 4, ...)
 * Entries with k%3==2 have odd predecessors (1, 3, 5, ...)
 *
 *  * Each entry has a sibling which is steps+1 from its successor away.
 */
abstract class Collatz {

    static final Comparator<Collatz> CMP = Comparator.comparing(Collatz::j).thenComparingInt(Collatz::m);

    public static Collatz max(Collatz a, Collatz b) {
        return CMP.compare(a, b) > 0 ? a : b;
    }

    static final BigInteger THREE = BigInteger.valueOf(3L);

    // k = 3*j + m;
    // n = 2*k + 1;
    final BigInteger j;
    //final int m;

    // index of this collatz
    //final BigInteger k;

    final int step;

    final int len;

    final Collatz succ;

    transient Collatz pred;

    final List<Collatz> seq;

    protected Collatz(Collatz succ, int step, BigInteger j) {
        this.succ = succ;
        this.len = succ.len+1;
        this.step = step;
        this.j = j;
        //this.m = m;
        //this.k = j.multiply(THREE).add(BigInteger.valueOf(m));

        this.seq = new AbstractList<>() {

            @Override
            public Collatz get(int index) {
                Collatz c = Collatz.this;

                for(int i=index+1; i<len; ++i)
                    c = c.succ;

                return c;
            }

            @Override
            public int size() {
                return len;
            }
        };
    }

    protected Collatz() {
        this.j = BigInteger.ZERO;
        this.step = 1;
        this.succ = this;
        this.pred = this;
        this.len = 0;
        this.seq = List.of(this);
    }

    @Override
    public String toString() {
        return String.format("3*%d+%d=%d[%d]", j, m(), k(), step);
    }

    abstract public int m();

    public BigInteger j() {
        return j;
    }

    public BigInteger k() {
        return j.multiply(THREE).add(BigInteger.valueOf(m()));
    }

    public Collatz succ() {
        return succ;
    }

    public Collatz pred(int i) {
        if(i<0)
            throw new IllegalArgumentException();

        if(i==0)
            return pred();
        
        Collatz pi = pred(i-1);
        return pi.sibl();
    }

    public Collatz pred() {
        return null;
    }

    transient Collatz sibl;

    public Collatz sibl() {
        Collatz sibl = this.sibl;

        if (sibl != null)
            return sibl;
        else
            return _sibl();
    }

    abstract protected Collatz _sibl();
}
